File: python/d07.py
# d07.py
# Advent of Code 2021 day 7 ver 1
# https://adventofcode.com/2021/day/7
# Treachery of whales: find optimal number of steps


def part1(input_file):
    dat = prep_data(input_file)
    fmin = len(dat)*max(dat)**2
    for i in range(max(dat)):
        fuel = sum([abs(t - i) for t in dat])
        fmin = min(fmin, fuel)

    return fmin


def part2(input_file):
    dat = prep_data(input_file)

    def recur_sum(n):
        return int((n + 1) * n / 2)

    fmin = len(dat)*max(dat)**2
    for i in range(max(dat)):
        fuel = sum([recur_sum(abs(t - i + 1)) for t in dat])
        fmin = min(fmin, fuel)

    return fmin


def read_lines(input_file):
    with open(input_file, "r") as fn:
        buf = fn.read()
    return [e.strip() for e in buf.split('\n') if len(e.strip()) > 0]


def prep_data(input_file):
    dat = read_lines(input_file)
    dat = list(map(int, dat[0].split(",")))
    return dat


if __name__ == "__main__":

    # validation and input files
    valid_file = "data/07a.txt"
    input_file = "data/07.txt"

    # part 1
    assert part1(valid_file) == 37
    assert part1(input_file) == 337833

    # part 2
    assert part2(valid_file) == 168
    assert part2(input_file) == 96678050

File: rust/src/d07.rs
// d06.rs

use crate::util;
use util::MyError;
use std::cmp;

pub fn part1(input_file: &str) -> Result<u32, MyError> {
    let data = parse_lines(util::read_lines(input_file));
    println!("data size {}", data.len());

    // initial search
    let mut fmin: u32 = (data.len() as u32) * 
        data.iter().max().unwrap().pow(2) as u32;

    for k in data.iter() {
        let fuel: i32 = data
            .iter()
            .map(|t: &i32| -> i32{(t - (*k as i32)).abs()})
            .collect::<Vec<i32>>()
            .iter()
            .sum();
        fmin = cmp::min(fmin, fuel as u32);
    }
    Ok(fmin)
}

pub fn part2(input_file: &str) -> Result<u32, MyError> {
    let data = parse_lines(util::read_lines(input_file));

    // initial search
    let mut fmin: u32 = (data.len() as u32) * 
        data.iter().max().unwrap().pow(2) as u32;

    // fuel cost function
    let recur_sum = |n: i32| (n + 1) * n / 2;

    for k in 0..*data.iter().max().unwrap() {
        let fuel: i32 = data
            .iter()
            .map(|t: &i32| recur_sum((t + 1 - (k as i32)).abs()))
            .collect::<Vec<i32>>()
            .iter()
            .sum();
        fmin = cmp::min(fmin, fuel as u32);
    }
    Ok(fmin)
}

fn parse_lines(lines: Vec<String>) -> Vec<i32> {
    lines[0]
        .split(',')
        .map(|t| t.parse::<i32>().expect("failed to parse integer"))
        .collect()
}
